ProjectLighthouse/Controllers/ReviewController.cs

```diff
@@ -74,5 +81,177 @@ public async Task<IActionResult> DPadRate(int slotId, [FromQuery] int rating)
             return this.Ok();
         }
 
+        [HttpPost("postReview/user/{slotId:int}")]
+        public async Task<IActionResult> PostReview(int slotId) {
+            User? user = await this.database.UserFromRequest(this.Request);
+            if (user == null) return this.StatusCode(403, "");
+
+            Review? review = await this.database.Reviews.FirstOrDefaultAsync(r => r.SlotId == slotId && r.ReviewerId == user.UserId);
+            Review? newReview = await this.GetReviewFromBody();
+            if (newReview == null) return this.BadRequest();
+            if (review == null) {
+                review = new();
+                review.SlotId = slotId;
+                review.ReviewerId = user.UserId;
+                review.DeletedBy = "none";
+            
+            }
+            review.LabelCollection = newReview.LabelCollection;
+            review.Text = newReview.Text;
+            review.Deleted = false;
+            review.Timestamp = TimeHelper.UnixTimeMilliseconds();
+            
+            // sometimes the game posts a review without also calling dpadrate/user/etc (why??)
+            RatedLevel? ratedLevel = await this.database.RatedLevels.FirstOrDefaultAsync(r => r.SlotId == slotId && r.UserId == user.UserId);
+            if (ratedLevel == null)
+            {
+                ratedLevel = new RatedLevel();
+                ratedLevel.SlotId = slotId;
+                ratedLevel.UserId = user.UserId;
+                ratedLevel.RatingLBP1 = 0;
+                this.database.RatedLevels.Add(ratedLevel);
+            }
+
+            ratedLevel.Rating = newReview.Thumb;
+            
+
+            await this.database.SaveChangesAsync();
+
+            return this.Ok();
+        }
+
+        [HttpGet("reviewsFor/user/{slotId:int}")]
+        public async Task<IActionResult> ReviewsFor(int slotId, [FromQuery] int pageStart = 1, [FromQuery] int pageSize = 10) 
+        {
+            User? user = await this.database.UserFromRequest(this.Request);
+            if (user == null) return this.StatusCode(403, "");
+
+            Token? token = await this.database.TokenFromRequest(this.Request);
+            if (token == null) return this.StatusCode(403, "");
+
+            GameVersion gameVersion = token.GameVersion;
+
+            Random rand = new();
+
+            IEnumerable<Review> reviews = this.database.Reviews.Where(r => r.SlotId == slotId && r.Slot.GameVersion <= gameVersion)
+                .Include(r => r.Reviewer)
+                .Include(r => r.Slot)
+                .AsEnumerable() // performance? Needed for next line (ThumbsUp is not in DB)
+                .OrderByDescending(r => r.ThumbsUp)
+                .ThenByDescending(_ => rand.Next())
```

Comment:
Hmm, it's a hack but we might be able to store and cache a list of reviews and their ThumbsUp values and send that to the DB that way. Otherwise we're left through ordering the entire table like this.